### ISSUE TITLE
Content Returned not being detected as json?

### ISSUE THREAD

**Issue by ravensorb** [canton7/RestEase#63]
Ok this might need some details.

I have an external API that I am calling (I don't have any control over it).  When I set the Header attribute for a "content-type: application/json" everything works expect this 1 single call. For this one call, when the header attribute exists, I get the following

```Header Content-Type could not be added. Maybe it's a content-related header but there's no content?```

When I remove the header attribute, I am able to make the call, however when I call "GetContent" I receive the following message

```You must override and implement T Deserialize<T>(string content, HttpResponseMessage response, ResponseDeserializerInfo info)```

Here is the raw response header (from fiddler)
```
HTTP/1.1 200 OK
Cache-Control: no-cache, no-store, max-age=0, must-revalidate, private
Content-Security-Policy: script-src 'self' 'unsafe-inline' 'unsafe-eval' data: https://www.google-analytics.com/analytics.js https://ajax.googleapis.com/ajax/libs/jquery/2.2.4/jquery.min.js https://usageanalytics.coveo.com/ https://code.jquery.com https://static.cloud.coveo.com
Content-Type: application/json;charset=utf-8
Date: Fri, 01 Sep 2017 13:54:03 GMT
Expires: 0
Pragma: no-cache
Strict-Transport-Security: max-age=31536000 ; includeSubDomains
Vary: Accept-Encoding, User-Agent
X-Content-Type-Options: nosniff
X-Frame-Options: deny
X-Request-ID: 608ad4f2-4622-4a4e-bd6c-1817b81bf849
X-XSS-Protection: 1; mode=block
Connection: keep-alive
Content-Length: 3235

[{
		"sourceType": "PUSH",
		"id": "1234",
		"name": "rm_index - DEV",
		"sourceVisibility": "SHARED",
		"information": {
			"sourceStatus": {
				"type": "PUSH_READY",
				"numberOfDocuments": 0,
				"allowedOperations": ["DELETE"]
			},
			"numberOfDocuments": 0,
			"documentsTotalSize": 0
		},
		"pushEnabled": true,
		"preConversionExtensions": [],
		"postConversionExtensions": [],
		"urlFilters": [{
				"filter": "*",
				"includeFilter": true,
				"filterType": "WILDCARD"
			}
		]
	}
]
```
As I said - it is only this call that is failing (I have about 15 others that are working 100% correctly with the Header attribute applied).

Any suggestions on what could be causing the issue?

The other interesting thing is even with AllowAnyStatusCode being set, I am still getting the first exception

**Comment by canton7** [comment]
There are two things going on here.

> Header Content-Type could not be added. Maybe it's a content-related header but there's no content?

HttpClient does not let you set the Content-Type header when the request does not have any content (that is, a request body). This means that you cannot have a Content-Type header attribute on the whole class, or a Content-Type property, unless all methods have a body.

You can work around this by putting the attribute `[Header("Content-Type", null)]` on the method which does not have a body. This will remove the Content-Type header for that method.

> You must override and implement T Deserialize<T>(string content, HttpResponseMessage response, ResponseDeserializerInfo info)

You will see this message if you have written a custom `ResponseDeserializer` and assigned it to `RestClient.ResponseDeserializer`, but you haven't overridden the method which lets you actually deserialize the response. Without seeing more of your code I can't say what exactly you're doing. Could you please post enough code to reproduce the issue?

**Comment by ravensorb** [comment]
Thanks for the quick response.  

So for the first option, I have tried that and it doesn't seem to work.  I have added ```[Header("Content-Type")]``` and ```[Header("Content-Type", null)]``` to the one method that doesn't have a Body -- both result in the same exception being thrown.  The only work around I have found is -- remove the Header from the Interface and add it to each method individually (a bit tedious however that got me past that specific issue).  Note: I am curious, why does it matter -- if there is no body, the content-type really doesn't matter, does it?

For the 2nd issue, I am not overriding the ResponseDeserializer right now (not in this project) and as I said, all other methods work, it is only the 1 method that is causing the issue

Here is one of the interfaces -- 3 of the 4 methods work.
```
    [AllowAnyStatusCode]
    public interface ICoveoContentSourceApi 
    {
        [Path("organizationId")]
        string OrganizationId { get; set; }

        #region Content Sources
        [Get("/{organizationId}/sources")]
        Task<Response<List<ContentSourceModel>>> GetSourcesAsync(); // this one doesn't work

        [Post("/{organizationId}/sources")]
        [Header("Content-Type", "application/json")]
        Task<Response<ContentSourceModel>> CreateSourceAsync([Body(BodySerializationMethod.Serialized)] ContentSourceModel contentSource, [Query("updateSecurityProviders")] bool updateSecurityProviders = false, [Query("rebuild")] bool rebuild = false);

        [Put("/{organizationId}/sources/{sourceId}")]
        Task<Response<ContentSourceModel>> UpdateSourceAsync([Path("sourceId")] string sourceId, [Body(BodySerializationMethod.Serialized)] ContentSourceModel contentSource);

        [Delete("/{organizationId}/sources/{sourceId}")]
        Task<Response<string>> DeleteSourceAsync([Path("sourceId")] string sourceId);
        #endregion Content Sources
    }
```

**Comment by canton7** [comment]
Aah yes I see the issue. It tries to add the Content-Type header first, before removing it. Except that it never gets as far as removing it, because it can't add it.

HttpClient stops you adding any headers to the request if they're associated with the content or response, and stops you adding any headers to the content if they're associated with the request or response. If the user has asked me to apply a header, I'm reluctant to silently drop it without any warning -- that's going to hide an issue. 

I've put a workaround in place.

For your second issue, there isn't enough code there for me to reproduce the issue. Please give me code that I can run which will reproduce your issue.

**Comment by ravensorb** [comment]
Thanks for working on the header issue :)

For the 2nd issue, I am working against a SAS solution Called Coveo -- https://platform.cloud.coveo.com/docs?api=Source

I'll see if I can great a small sample -- the only issue is, I cannot share the organizational key to test it out.  Is there anything else I can provide if the interface isn't enough?

**Comment by canton7** [comment]
You should be able to reproduce the issue against any API - RestEase doesn't do detection of the returned content type. 

Seeing the stack trace would also help.

**Comment by canton7** [comment]
Hi, any more input on this?

**Comment by ravensorb** [comment]
Is this what you are looking for?

```
   at RestEase.ResponseDeserializer.Deserialize[T](String content, HttpResponseMessage response, ResponseDeserializerInfo info)
   at RestEase.Response`1.GetContent()
   at Coveo.Tests.CoveFieldApi.<GetFieldsAsyncTest>d__3.MoveNext() in E:\POC\CoveoDBIndexer\tests\Coveo.Tests\CoveFieldApi.Tests.cs:line 60
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Coveo.Tests.CoveFieldApi.<FieldsTests>d__2.MoveNext() in E:\POC\CoveoDBIndexer\tests\Coveo.Tests\CoveFieldApi.Tests.cs:line 44
```

**Comment by canton7** [comment]
That's really weird, and I still can't reproduce this.

```csharp
    public class User
    {
        public string Name { get; set; }
    }

    [Header("User-Agent", "RestEase")]
    public interface IGitHubApi
    {
        [Get("users/{userId}")]
        Task<Response<User>> GetUserAsync([Path] string userId);
    }

    class Program
    {
        
        public static void Main(string[] args)
        {
            IGitHubApi api = RestClient.For<IGitHubApi>("https://api.github.com");

            var response = api.GetUserAsync("canton7").Result;
            var user = response.GetContent();
            Console.WriteLine($"Name: {user.Name}");
            Console.ReadLine();
        }
    }
```

The only possible explanation I can think of is that you're using a custom ResponseDeserializer but haven't overridden the correct method.

**Comment by canton7** [comment]
v1.4.4 improves that exception message. Mind updating and giving me the new exception message?

**Comment by ravensorb** [comment]
Just got a chance to look back at this.  From the Content-Type issue I am still getting it even in the new build

```
Header Content-Type could not be added. Maybe it's a content-related header but there's no content?
```
```
   at RestEase.Implementation.Requester.ApplyHeadersSet(HttpRequestMessage requestMessage, IEnumerable`1 headers)
   at RestEase.Implementation.Requester.ApplyHeaders(IRequestInfo requestInfo, HttpRequestMessage requestMessage)
   at RestEase.Implementation.Requester.<SendRequestAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at RestEase.Implementation.Requester.<RequestWithResponseAsync>d__33`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Program.<ExecuteProcessAsync>d__1.MoveNext() in E:\Projects\POC\Program.cs:line 73
``

**Comment by canton7** [comment]
This message: 

> Header Content-Type could not be added. Maybe it's a content-related header but there's no content?

Is not present in v1.4.4. The string was changed. Therefore you are not running v1.4.4.

**Comment by soarartsdev** [comment]
I just experienced this same issue as well. An example is PayPal's "order capture" endpoint (see [documentation](https://developer.paypal.com/docs/api/orders/v2/#orders_capture))

PayPal _requires_ a `Content-Type` header even though that POST endpoint takes in only one path parameter...

So my workaround was to explicitly pass an empty request body. Goofy, right?

Here's a small repro:
```
using System.Collections.Generic;
using System.Threading.Tasks;
using Models.PayPalModels;

namespace Services
{
    public interface IPaymentService
    {
        Task<CaptureOrderResponse> CapturePaymentOrder(
            string encodedCredentials,
            string orderID
        );
    }

    public class PaymentService : IPaymentService
    {
        private readonly IPayPalApi _payPalApi;

        public PaymentService(IPayPalApi payPalApi)
        {
            _payPalApi = payPalApi;
        }

        public async Task<CaptureOrderResponse> CapturePaymentOrder(
            string encodedCredentials,
            string orderID
        )
        {
            var tokenResponse = await GetAccessTokenAsync(encodedCredentials);
            _payPalApi.AuthorizationHeader = $"Bearer {tokenResponse.AccessToken}";

            var result = await _payPalApi.CaptureOrderAsync(orderID);
            return result;
        }

        private async Task<AccessTokenResponse> GetAccessTokenAsync(string encodedCredentials)
        {
            _payPalApi.AuthorizationHeader = $"Basic {encodedCredentials}";
            var formData = new Dictionary<string, string>
            {
                { "grant_type", "client_credentials" }
            };
            return await _payPalApi.GetTokenAsync(formData);
        }
    }
}

```

And IPayPalApi:
```
using System.Collections.Generic;
using System.Threading.Tasks;
using Models.PayPalModels;
using RestEase;

namespace Services
{
    public interface IPayPalApi
    {
        [Header("Authorization")]
        string AuthorizationHeader { get; set; }

        [Post("v1/oauth2/token")]
        [Header("Accept", "application/json")]
        [Header("Accept-Language", "en_US")]
        [Header("Content-Type", "application/x-www-form-urlencoded")]
        Task<AccessTokenResponse> GetTokenAsync([Body(BodySerializationMethod.UrlEncoded)] Dictionary<string, string> formData);

        [Post("v2/checkout/orders/{orderID}/capture")]
        [Header("Content-Type", "application/json")] // without this, PayPal returns UNSUPPORTED_MEDIA_TYPE response
        Task<CaptureOrderResponse> CaptureOrderAsync([Path] string orderID, [Body] object _ = null); // workaround is to pass an empty body
    }
}

```

**Comment by canton7** [comment]
What version are you using?

**Comment by soarartsdev** [comment]
1.4.12

**Comment by canton7** [comment]
Right, so the behaviour is that we add dummy content only if there's a content-related header on the interface itself, not on the method. This is to fix #109 

Maybe there's an argument saying we should always add a dummy body. Can you open a new issue?
